Make thrustcurve selection clearer for users

JoePfeiffer opened the pull request:
One of our not infrequent complaints from users is that a motor in the database has an inaccurate thrustcurve, or that its thrustcurve has changed for the worse. This is typically because our motor selection dialog fails to make clear to the user that a motor may have multiple thrustcurves to choose from. This PR attempts to address this.

At present, when a motor is selected in the dialog, the only indication that the user has a choice of thrustcurves is the presence of the "Select thrust curve" combobox in the upper left corner of the dialog.  Notably the "Filter Motors" pane remains selected, as shown in the following screenshot:

<img width="911" height="734" alt="image" src="https://github.com/user-attachments/assets/fce738ce-abd2-4cc5-88e8-eda4eb3f75c8" />

In order to see the motor details, the user must select the "Show Details" pane:

<img width="911" height="734" alt="image" src="https://github.com/user-attachments/assets/8474d400-8756-4cc5-8af5-f04deac172ca" />

Even now, the thrustcurve selection combobox is literally as far away from the thrustcurve plot as it can possibly get and be in the same dialog. Also, the label in the combobox doesn't reflect the selected thrustcurve, as it's black (through version 23.09, the color matched the color of the curve; in 24.12 it does not. I'm actually not sure how that changed; I haven't found where in the code it was set to match in the old versions).

This PR attempts to improve this situation. First, when a motor is selected, the motor detail pane is selected automatically (the user can switch back to the filter pane either by selecting the pane, or by entering text in the search field). The thrust curve selection combobox, "hide similar motors" checkbox, and ejection charge delay are all moved to the detail pane. The thrust curve plot is moved to just under the comboboxes, and the color of the thrust curve selection combobox matches the color of the currently selected thrustcurve.

<img width="911" height="843" alt="image" src="https://github.com/user-attachments/assets/e7b753a5-c9b6-4278-99ff-eb1649703de6" />

JoePfeiffer committed:
When creating a new configuration, don't start with a preselected motor

JoePfeiffer committed:
When opening motor selection dialog, clear search field
Clear motor selection when typing in search field

JoePfeiffer committed:
When current motor selection is cleared, bring up motor filter pane
When a motor is selected, bring up motor detail pane

The user is still able to switch panes manually, so (for instance) while a motor is selected the motor filter panel can be brought up without clearing the motor selection

JoePfeiffer committed:
Add iterator and other necessary methods to ThrustCurveMotorSetDatabase, and use it for database operations instead of reimplementing them in ThrustCurveMotorSelectionPanel (I wonder if there are other places this should be done?)

JoePfeiffer committed:
Move thrust curve display to top of MotorInformationPanel

JoePfeiffer committed:
add indexOf() method to get index of motor in ArrayList implementing ThrustCurveMotorSet

JoePfeiffer committed:
only use table to select motor sets, not motors

JoePfeiffer committed:
A much larger commit than I like to make.

Move elements of ThrustCurveMotorSelectionPanel relating to selecting a specific thrustcurve and setting ejection charge delay to the MotorInformationPanel.

Specifically, move
    ejection charge delay combobox
    hide similar thrust curve
    select thrust curve combobox

Also move thrust curve plot from bottom of pane to just under thrust curve selection combobox

JoePfeiffer committed:
Set thrustcurve selection combobox foreground color to match color of corresponding thrust curve.

JoePfeiffer committed:
Merge branch 'unstable' into motor-ui

neilweinstock commented:
I had to stare at it a bit but I think I like it.

I'm wondering if there's a way to improve the delay selection, while we're at it.  I'm thinking a visible list of available delays with radio buttons, plus a blank custom delay field that is also selectable.  The current combobox fools a lot of people.

I'll try to throw together a mockup, or maybe there's another better way to do it.

jimmiedave commented:
I generally like it. I'd like to see a color chip, vs. color text in the curve selection droplist. More color onscreen to assist in discernment. The color text is subtle for me.

<img width="292" height="217" alt="Screenshot 2025-09-30 at 8 39 09 PM" src="https://github.com/user-attachments/assets/31e82529-ee98-4ac7-8786-6ebd623826c8" />

thzero commented:
I'm not a fan of moving the ejection charge to under the Show Details as it is now.

If however when I highlight a motor the Show Details is automatically switched, then cool.... that makes sense.  The ejection charge, the different thrust curves, etc. is all right there.

I like the chip color, but for people who are color blind, that makes it hard.  

Still prefer dropdown with custom text box as opposed to radio buttons.

JoePfeiffer commented:
This is all why I requested comments!

Yes, the color chip makes the difference clearer -- I want to do that. Using color at all causes issues for color blind people (my father was red-green color blind. I've got stories....), but I don't have a better idea here.

I really like the radio buttons for delay (sorry @thzero). It seems like a clearer way of letting the user know their options.

I put the "hide similar" at the top of that section for consistency with "hide OOP", Not important?

Yes, when you select a motor it automatically switches to the motor detail pane. While I like all the other changes (I'd better, since I did most of them), that automatic switch is the single most important one.

neilweinstock commented:
> I put the "hide similar" at the top of that section for consistency with "hide OOP", Not important?

No, not in my opinion.  Once consistency starts making things worse then it gets tossed out the window.  I would venture to guess that use of the "hide similar" checkbox is extraordinarily rare, it should be moved out of the way of what they're really interested in.  I would probably also prefer that the "hide out-of-production" box also go below the table, but I don't think it's causing much harm being up top.

> Yes, when you select a motor it automatically switches to the motor detail pane. While I like all the other changes (I'd better, since I did most of them), that automatic switch is the single most important one.

It is fundamental to this whole PR; without it you could not possibly hide the controls in question in the details tab.

JoePfeiffer committed:
Instead of using the plot color for the text on a motor thrustcurve, leave the foreground color alone and use a color chip with the same color as the thrustcurve.

JoePfeiffer committed:
Merge branch 'unstable' into motor-ui

JoePfeiffer committed:
Merge branch 'unstable' into motor-ui

JoePfeiffer committed:
Merge branch 'unstable' into motor-ui

JoePfeiffer committed:
Remove unneeded chart title
Put comments right under thrustcurve chart
Move "similar curves" checkbox under chart

JoePfeiffer committed:
wip
